daffodil-lib/src/main/scala/edu/illinois/ncsa/daffodil/util/DecimalUtils.scala

```diff
@@ -0,0 +1,328 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package edu.illinois.ncsa.daffodil.util
+
+import edu.illinois.ncsa.daffodil.exceptions.Assert
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.BinaryNumberCheckPolicy
+
+import java.math.{ BigInteger => JBigInteger, BigDecimal => JBigDecimal }
+
+object PackedSignCodes {
+  def apply(signCodes: String, policy: BinaryNumberCheckPolicy) = {
+    val str = signCodes.replaceAll("\\s", "")
+    val chars = str.toCharArray()
+    Assert.invariant(chars.length == 4)
+
+    val positive = policy match {
+      case BinaryNumberCheckPolicy.Strict => List(Integer.parseInt(String.valueOf(chars(0)), 16))
+      case BinaryNumberCheckPolicy.Lax => List(Integer.parseInt(String.valueOf(chars(0)), 16), 0xA, 0xC, 0xE, 0xF)
+    }
+    val negative = policy match {
+      case BinaryNumberCheckPolicy.Strict => List(Integer.parseInt(String.valueOf(chars(1)), 16))
+      case BinaryNumberCheckPolicy.Lax => List(Integer.parseInt(String.valueOf(chars(1)), 16), 0xB, 0xD)
+    }
+    val unsigned = policy match {
+      case BinaryNumberCheckPolicy.Strict => List(Integer.parseInt(String.valueOf(chars(2)), 16))
+      case BinaryNumberCheckPolicy.Lax => List(Integer.parseInt(String.valueOf(chars(2)), 16), 0xF)
+    }
+    val zero_sign = policy match {
+      case BinaryNumberCheckPolicy.Strict => List(Integer.parseInt(String.valueOf(chars(3)), 16))
+      case BinaryNumberCheckPolicy.Lax => List(Integer.parseInt(String.valueOf(chars(3)), 16), 0xA, 0xC, 0xE, 0xF, 0x0)
+    }
+
+    new PackedSignCodes(positive, negative, unsigned, zero_sign)
+  }
+}
+
+case class PackedSignCodes(
+  positive: List[Int],
+  negative: List[Int],
+  unsigned: List[Int],
+  zero_sign: List[Int]
+) {}
+
+object DecimalUtils {
+
+  def packedToBigInteger(num: Array[Byte], signCodes: String, policy: BinaryNumberCheckPolicy): JBigInteger = {
```

Comment:
This function and similar *ToBigInteger functions are called at parse/unpars time, so it will also call PackedSignCodes.apply(...) at runtime. But we have all the information necessary (signCodes and binaryNumberCheckPolicy) at schema compile time. So we should be able to move the  PackedSignCodes.apply(...) call to somewhere in schema compile time (perhaps a lazy val in the grammar and then passed into the *Parser/Unparser constructors instead of signCode and policy), we can minimize the performance cost at runtime of building this objects. Then these functions could all take a PackedSignCodes object instead of the signCodes String and binaryNumberCheckPolicy.

Also, PackedSignCodes(...) without the 'new' keyword will cal the apply method of the object class, so .apply(...) isn't necessary and is usually left off in Scala.

Reply:
Will do